File: examples/deno/apps/agent/mcp_stdio.ts
import { TextLineStream } from "@std/streams/text-line-stream";

import {
  type AppMcpCallResult,
  type AppMcpInitializeResult,
  AppMcpProtocolError,
  type AppMcpReadResourceResult,
  type AppMcpRequestContext,
  type AppMcpResourceDescriptor,
  type AppMcpServer,
  type AppMcpToolDescriptor,
  BaseAppMcpProcessAdapter,
} from "../../core/shared/app_mcp_contracts";

type JsonRpcId = string | number | null;

type JsonRpcRequest = {
  jsonrpc: "2.0";
  id: JsonRpcId;
  method: string;
  params?: unknown;
};

type JsonRpcNotification = {
  jsonrpc: "2.0";
  method: string;
  params?: unknown;
};

type JsonRpcSuccess = {
  jsonrpc: "2.0";
  id: JsonRpcId;
  result:
    | AppMcpInitializeResult
    | { tools: AppMcpToolDescriptor[] }
    | { resources: AppMcpResourceDescriptor[] }
    | AppMcpReadResourceResult
    | AppMcpCallResult
    | Record<string, never>;
};

type JsonRpcFailure = {
  jsonrpc: "2.0";
  id: JsonRpcId;
  error: {
    code: number;
    message: string;
    data?: unknown;
  };
};

type SessionPhase =
  | "awaiting_initialize"
  | "awaiting_initialized_notification"
  | "ready";

const encoder = new TextEncoder();

function generateId(): string {
  return globalThis.crypto?.randomUUID?.() ??
    Math.random().toString(36).slice(2);
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function toFailure(
  id: JsonRpcId,
  code: number,
  message: string,
  data?: unknown,
): JsonRpcFailure {
  return {
    jsonrpc: "2.0",
    id,
    error: {
      code,
      message,
      ...(data !== undefined ? { data } : {}),
    },
  };
}

function isJsonRpcMessage(value: unknown): value is {
  jsonrpc: "2.0";
  method: string;
  id?: JsonRpcId;
  params?: unknown;
} {
  return (
    isObject(value) &&
    value.jsonrpc === "2.0" &&
    typeof value.method === "string"
  );
}

function isInitializeParams(value: unknown): value is {
  protocolVersion: string;
  capabilities?: Record<string, unknown>;
  clientInfo?: {
    name: string;
    version: string;
  };
} {
  if (!isObject(value) || typeof value.protocolVersion !== "string") {
    return false;
  }

  if (value.clientInfo === undefined) {
    return true;
  }

  return (
    isObject(value.clientInfo) &&
    typeof value.clientInfo.name === "string" &&
    typeof value.clientInfo.version === "string"
  );
}

function writeMessage(message: JsonRpcSuccess | JsonRpcFailure): void {
  Deno.stdout.writeSync(encoder.encode(`${JSON.stringify(message)}\n`));
}

export class StdioAppMcpAdapter extends BaseAppMcpProcessAdapter {
  public readonly transport = "stdio" as const;

  public async serve(server: AppMcpServer): Promise<void> {
    const sessionId = generateId();
    let phase: SessionPhase = "awaiting_initialize";
    let protocolVersion: string | undefined;
    let clientInfo:
      | {
        name: string;
        version: string;
      }
      | undefined;

    const handleRequest = async (
      request: JsonRpcRequest,
      context: AppMcpRequestContext,
    ): Promise<JsonRpcSuccess> => {
      switch (request.method) {
        case "initialize": {
          if (phase !== "awaiting_initialize") {
            throw new AppMcpProtocolError(
              -32600,
              "MCP initialize may only run once per stdio session.",
            );
          }

          const result = await server.initialize(
            isInitializeParams(request.params) ? request.params : undefined,
            context,
          );
          protocolVersion = result.protocolVersion;
          clientInfo =
            isObject(request.params) && isObject(request.params.clientInfo)
              ? {
                name: typeof request.params.clientInfo.name === "string"
                  ? request.params.clientInfo.name
                  : "unknown-client",
                version: typeof request.params.clientInfo.version === "string"
                  ? request.params.clientInfo.version
                  : "0.0.0",
              }
              : undefined;
          phase = "awaiting_initialized_notification";

          return {
            jsonrpc: "2.0",
            id: request.id,
            result,
          };
        }
        case "ping":
          if (phase !== "ready") {
            throw new AppMcpProtocolError(
              -32002,
              "MCP session is not ready; complete initialization first.",
            );
          }

          return {
            jsonrpc: "2.0",
            id: request.id,
            result: {},
          };
        case "tools/list":
          if (phase !== "ready") {
            throw new AppMcpProtocolError(
              -32002,
              "MCP session is not ready; complete initialization first.",
            );
          }

          return {
            jsonrpc: "2.0",
            id: request.id,
            result: {
              tools: await server.listTools(context),
            },
          };
        case "resources/list":
          if (phase !== "ready") {
            throw new AppMcpProtocolError(
              -32002,
              "MCP session is not ready; complete initialization first.",
            );
          }

          return {
            jsonrpc: "2.0",
            id: request.id,
            result: {
              resources: await server.listResources(context),
            },
          };
        case "resources/read":
          if (phase !== "ready") {
            throw new AppMcpProtocolError(
              -32002,
              "MCP session is not ready; complete initialization first.",
            );
          }

          if (
            !isObject(request.params) || typeof request.params.uri !== "string"
          ) {
            throw new AppMcpProtocolError(
              -32602,
              "MCP resources/read requires a string resource uri.",
            );
          }

          return {
            jsonrpc: "2.0",
            id: request.id,
            result: await server.readResource(request.params.uri, context),
          };
        case "tools/call":
          if (phase !== "ready") {
            throw new AppMcpProtocolError(
              -32002,
              "MCP session is not ready; complete initialization first.",
            );
          }

          if (
            !isObject(request.params) || typeof request.params.name !== "string"
          ) {
            throw new AppMcpProtocolError(
              -32602,
              "MCP tools/call requires a string tool name.",
            );
          }

          return {
            jsonrpc: "2.0",
            id: request.id,
            result: await server.callTool(
              request.params.name,
              request.params.arguments,
              context,
            ),
          };
        default:
          throw new AppMcpProtocolError(
            -32601,
            `MCP method ${request.method} is not implemented by this server.`,
          );
      }
    };

    const handleNotification = (notification: JsonRpcNotification): void => {
      switch (notification.method) {
        case "notifications/initialized":
          if (phase === "awaiting_initialized_notification") {
            phase = "ready";
          }
          return;
        case "notifications/cancelled":
          return;
        default:
          return;
      }
    };

    const toContext = (requestId?: JsonRpcId): AppMcpRequestContext => ({
      transport: this.transport,
      requestId,
      sessionId,
      correlationId: generateId(),
      clientInfo,
      protocolVersion,
    });

    const lines = Deno.stdin.readable
      .pipeThrough(new TextDecoderStream())
      .pipeThrough(new TextLineStream());

    for await (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) {
        continue;
      }

      let parsed: unknown;
      try {
        parsed = JSON.parse(trimmed) as unknown;
      } catch (error: unknown) {
        writeMessage(
          toFailure(
            null,
            -32700,
            "Invalid JSON-RPC payload.",
            error instanceof Error ? error.message : undefined,
          ),
        );
        continue;
      }

      if (!isJsonRpcMessage(parsed)) {
        writeMessage(
          toFailure(null, -32600, "Invalid JSON-RPC request shape."),
        );
        continue;
      }

      const hasId = Object.prototype.hasOwnProperty.call(parsed, "id");
      if (!hasId) {
        handleNotification(parsed);
        continue;
      }

      const request: JsonRpcRequest = {
        jsonrpc: "2.0",
        id: typeof parsed.id === "string" || typeof parsed.id === "number"
          ? parsed.id
          : null,
        method: parsed.method,
        params: parsed.params,
      };

      try {
        const response = await handleRequest(request, toContext(request.id));
        writeMessage(response);
      } catch (error: unknown) {
        if (error instanceof AppMcpProtocolError) {
          writeMessage(
            toFailure(request.id, error.code, error.message, error.data),
          );
          continue;
        }

        writeMessage(
          toFailure(
            request.id,
            -32603,
            "Internal MCP server error.",
            error instanceof Error ? { message: error.message } : undefined,
          ),
        );
      }
    }
  }
}

File: examples/flutter/apps/agent/mcp_stdio.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import '../../core/shared/app_mcp_contracts.dart';

typedef _JsonRpcId = Object?;

Map<String, dynamic> _toStdioFailure(
  _JsonRpcId id,
  int code,
  String message, [
  dynamic data,
]) {
  return <String, dynamic>{
    'jsonrpc': '2.0',
    'id': id,
    'error': <String, dynamic>{
      'code': code,
      'message': message,
      if (data != null) 'data': data,
    },
  };
}

bool _isStdioJsonRpc(dynamic value) {
  return value is Map<String, dynamic> &&
      value['jsonrpc'] == '2.0' &&
      value['method'] is String;
}

AppMcpInitializeParams? _parseStdioInitializeParams(dynamic value) {
  if (value is! Map<String, dynamic> || value['protocolVersion'] is! String) {
    return null;
  }

  final clientInfoValue = value['clientInfo'];
  AppMcpClientInfo? clientInfo;
  if (clientInfoValue is Map<String, dynamic> &&
      clientInfoValue['name'] is String &&
      clientInfoValue['version'] is String) {
    clientInfo = AppMcpClientInfo(
      name: clientInfoValue['name'] as String,
      version: clientInfoValue['version'] as String,
    );
  }

  return AppMcpInitializeParams(
    protocolVersion: value['protocolVersion'] as String,
    capabilities: value['capabilities'] is Map<String, dynamic>
        ? Map<String, dynamic>.from(
            value['capabilities'] as Map<String, dynamic>,
          )
        : null,
    clientInfo: clientInfo,
  );
}

class StdioAppMcpAdapter extends BaseAppMcpProcessAdapter {
  @override
  String get transport => 'stdio';

  @override
  Future<void> serve(AppMcpServer server) async {
    final sessionId = 'stdio_${DateTime.now().microsecondsSinceEpoch}';
    var phase = 'awaiting_initialize';
    String? protocolVersion;
    AppMcpClientInfo? clientInfo;

    final completer = Completer<void>();
    late final StreamSubscription<String> subscription;

    Future<void> writeMessage(Map<String, dynamic> message) async {
      stdout.writeln(jsonEncode(message));
    }

    subscription = stdin
        .transform(utf8.decoder)
        .transform(const LineSplitter())
        .listen(
          (String line) async {
            final trimmed = line.trim();
            if (trimmed.isEmpty) {
              return;
            }

            dynamic parsed;
            try {
              parsed = jsonDecode(trimmed);
            } catch (error) {
              await writeMessage(
                _toStdioFailure(
                  null,
                  -32700,
                  'Invalid JSON-RPC payload.',
                  error.toString(),
                ),
              );
              return;
            }

            if (!_isStdioJsonRpc(parsed)) {
              await writeMessage(
                _toStdioFailure(
                  null,
                  -32600,
                  'Invalid JSON-RPC request shape.',
                ),
              );
              return;
            }

            final message = Map<String, dynamic>.from(parsed as Map);
            final hasId = message.containsKey('id');
            final requestId = message['id'];
            final method = message['method'] as String;
            final context = AppMcpRequestContext(
              transport: transport,
              requestId: requestId,
              sessionId: sessionId,
              protocolVersion: protocolVersion,
              clientInfo: clientInfo,
            );

            if (!hasId) {
              if (method == 'notifications/initialized' &&
                  phase == 'awaiting_initialized_notification') {
                phase = 'ready';
              }
              return;
            }

            try {
              switch (method) {
                case 'initialize':
                  if (phase != 'awaiting_initialize') {
                    throw const AppMcpProtocolError(
                      -32600,
                      'MCP initialize may only run once per stdio session.',
                    );
                  }

                  final params = _parseStdioInitializeParams(message['params']);
                  final result = await server.initialize(params, context);
                  protocolVersion = result.protocolVersion;
                  clientInfo = params?.clientInfo;
                  phase = 'awaiting_initialized_notification';

                  await writeMessage(<String, dynamic>{
                    'jsonrpc': '2.0',
                    'id': requestId,
                    'result': result.toJson(),
                  });
                  return;
                case 'ping':
                  if (phase != 'ready') {
                    throw const AppMcpProtocolError(
                      -32002,
                      'MCP session is not ready; complete initialization first.',
                    );
                  }

                  await writeMessage(<String, dynamic>{
                    'jsonrpc': '2.0',
                    'id': requestId,
                    'result': <String, dynamic>{},
                  });
                  return;
                case 'tools/list':
                  if (phase != 'ready') {
                    throw const AppMcpProtocolError(
                      -32002,
                      'MCP session is not ready; complete initialization first.',
                    );
                  }

                  await writeMessage(<String, dynamic>{
                    'jsonrpc': '2.0',
                    'id': requestId,
                    'result': <String, dynamic>{
                      'tools': (await server.listTools(context))
                          .map((AppMcpToolDescriptor item) => item.toJson())
                          .toList(growable: false),
                    },
                  });
                  return;
                case 'resources/list':
                  if (phase != 'ready') {
                    throw const AppMcpProtocolError(
                      -32002,
                      'MCP session is not ready; complete initialization first.',
                    );
                  }

                  await writeMessage(<String, dynamic>{
                    'jsonrpc': '2.0',
                    'id': requestId,
                    'result': <String, dynamic>{
                      'resources': (await server.listResources(context))
                          .map((AppMcpResourceDescriptor item) => item.toJson())
                          .toList(growable: false),
                    },
                  });
                  return;
                case 'resources/read':
                  if (phase != 'ready') {
                    throw const AppMcpProtocolError(
                      -32002,
                      'MCP session is not ready; complete initialization first.',
                    );
                  }

                  final params = message['params'];
                  if (params is! Map<String, dynamic> ||
                      params['uri'] is! String) {
                    throw const AppMcpProtocolError(
                      -32602,
                      'MCP resources/read requires a string resource uri.',
                    );
                  }

                  await writeMessage(<String, dynamic>{
                    'jsonrpc': '2.0',
                    'id': requestId,
                    'result': (await server.readResource(
                      params['uri'] as String,
                      context,
                    )).toJson(),
                  });
                  return;
                case 'tools/call':
                  if (phase != 'ready') {
                    throw const AppMcpProtocolError(
                      -32002,
                      'MCP session is not ready; complete initialization first.',
                    );
                  }

                  final params = message['params'];
                  if (params is! Map<String, dynamic> ||
                      params['name'] is! String) {
                    throw const AppMcpProtocolError(
                      -32602,
                      'MCP tools/call requires a string tool name.',
                    );
                  }

                  await writeMessage(<String, dynamic>{
                    'jsonrpc': '2.0',
                    'id': requestId,
                    'result': (await server.callTool(
                      params['name'] as String,
                      params['arguments'],
                      context,
                    )).toJson(),
                  });
                  return;
                default:
                  throw AppMcpProtocolError(
                    -32601,
                    'Unsupported MCP JSON-RPC method $method.',
                  );
              }
            } on AppMcpProtocolError catch (error) {
              await writeMessage(
                _toStdioFailure(
                  requestId,
                  error.code,
                  error.message,
                  error.data,
                ),
              );
            } catch (error) {
              await writeMessage(
                _toStdioFailure(
                  requestId,
                  -32603,
                  'Internal MCP server error.',
                  <String, dynamic>{'message': error.toString()},
                ),
              );
            }
          },
          onDone: () {
            if (!completer.isCompleted) {
              completer.complete();
            }
          },
          onError: (Object error, StackTrace stackTrace) {
            stderr.writeln('[mcp-stdio] $error');
            if (!completer.isCompleted) {
              completer.completeError(error, stackTrace);
            }
          },
        );

    await completer.future;
    await subscription.cancel();
  }
}
